source/cluster_manager/clusterManager.js

```diff
@@ -555,204 +566,945 @@ var ClusterManager = function (clusterName, selfId, spec) {
+        }catch(e){
+            log.fatal(util.inspect(e));
+        }finally {
+            let end = (new Date()).getTime();
+            let delay = end-begin;
+            if(delay>150){
+                log.debug(state.id, `follower.verifyLeader handler delay:`, `${delay}ms`,"leaderId:", state.leaderId, "self:", state);
+            }
+            mutex.unlock();
+        }
+    });
+
+    var changeRole = async function(newRole,newTerm){
+        if (state.role != "follower") {
+            log.warn('follower cycle in changeRole', "self:",state);
+            return;
+        }
+
+        state.leaderId = 0;
+        state.term = newTerm;
+        state.role = newRole;
+        log.info(`follower->${newRole}`,"self:",state);
+        await topicChannel.unsubscribe(routerKeys);
+        switch(newRole){
+            case "candidate":
+                await runAsCandidate(topicChannel, manager);
+                break;
+            case "follower":
+                await runAsFollower(topicChannel, manager);
+                break;
+
+        }
+    }
+
+    var installSnapshot = async function (data) {
+        if (!isInstallSnapshot) {
+            isInstallSnapshot = true;
+            log.info("start installSnapshot last commitId:",state.commitId,"data.commitId:",data.commitId,"from:",data.id);
+            topicChannel.publish(`clusterManager.leader.${state.leaderId}`, {type: 'installSnapshot', data: state});
+            return new Promise((resolve) => {
+                let count = 0;
+                let checkIntvl = setInterval(() => {
+                    if (!isInstallSnapshot) {
+                        resolve(true);
+                    } else {
+                        if ((++count) == 25) {
+                            isInstallSnapshot = false;
+                            clearInterval(checkIntvl);
+                            checkIntvl = null;
+                            resolve(false);
+                        }
+                    }
+                }, 20);
+            });
         }
     };
 
-    var superviseMaster = function () {
-        interval && clearInterval(interval);
-        interval = undefined;
-        if( role !== "slave"){
-            log.warn('cycle in slave change role:', role);
+    var verifyLeader = async function () {
+        return new Promise((resolve) => {
+            log.info("verifyLeader begin:", "self:", state);
+            if (state.leaderId == 0) {
+                resolve(false);
+                return;
+            }
+            if(!isVerifyLeader){
+                isVerifyLeader = true;
+                verifyLeaderResult = false;
+                topicChannel.publish(`clusterManager.leader.${state.leaderId}`, {type: 'verifyLeader', data: state});
+            }
+
+            verifyLeaderCallback.push(resolve);
+            let step = 20;
+            let count = getTimerCount(100,step);
+            let checkIntvl = setInterval(() => {
```

Comment:
Is it similar as installSnapshot check?
If you can use a callback instead of checking it by setInterval, you can get result as soon as it finishes.

Reply:
> Is it similar as installSnapshot check? If you can use a callback instead of checking it by setInterval, you can get result as soon as it finishes.

ok